com.unity.netcode.gameobjects/Runtime/Serialization/IBufferSerializerImplementation.cs

```diff
@@ -0,0 +1,37 @@
+using System;
+using UnityEngine;
+
+namespace Unity.Netcode
+{
+    public interface IBufferSerializerImplementation
+    {
+        bool IsReader { get; }
+        bool IsWriter { get; }
+
+        ref FastBufferReader GetFastBufferReader();
```

Comment:
why do we want to **force** implementors to have these (reader/writer) on their API?

Reply:
What would you recommend? Every time we discussed this in the past, what I heard was "give them the easy and safe option, but give them a way to turn it into the difficult but more flexible/efficient one". How would you propose exposing FastBufferReader and FastBufferWriter without having methods that can expose them?

One thing to say here is that this isn't intended to be something users can provide their own implementations of... really should be internal. It's meant to support only and exactly two implementations.

Reply:
a dev can construct a FastBufferReader, then use a method call to say: "hey here's the buffer you can use" and they will be using that and interface would still be clean, no?

Reply:
maybe I'm not understanding the relation/dependency between reader/writer and buffer?

Reply:
BufferSerializer is a thin wrapper around reader/writer to make them bi-directional. But checking IsReader and IsWriter isn't the purpose of this function... it's "I know what I'm doing and I would like to work with FastBufferReader, not BufferSerializer, because FastBufferReader gives me options BufferSerializer doesn't".

BufferSerializer also isn't intended to be constructed by the user... it's only used for INetworkSerializable.